src/Analysis/Ast/Test/LintNoSelfArgumentTests.cs

```diff
@@ -0,0 +1,173 @@
+
+        [TestCleanup]
+        public void Cleanup() => TestEnvironmentImpl.TestCleanup();
+
+        [TestMethod, Priority(0)]
+        public async Task FirstArgumentMethodNotSelf() {
+            const string code = @"
+class Test:
+    def test(x, y, z):
+        pass
+";
+            var analysis = await GetAnalysisAsync(code);
+            analysis.Diagnostics.Should().HaveCount(1);
+
+            var diagnostic = analysis.Diagnostics.ElementAt(0);
+            diagnostic.ErrorCode.Should().Be(ErrorCodes.NoSelfArgument);
+            diagnostic.SourceSpan.Should().Be(3, 9, 3, 13);
+            diagnostic.Message.Should().Be(Resources.NoSelfArgument);
+        }
+
+        [TestMethod, Priority(0)]
+        public async Task FirstArgumentPropertyNotSelf() {
+            const string code = @"
+class Test:
+    @property
+    def test(x, y, z):
+        pass
+";
+            var analysis = await GetAnalysisAsync(code);
+            analysis.Diagnostics.Should().HaveCount(1);
+
+            var diagnostic = analysis.Diagnostics.ElementAt(0);
+            diagnostic.ErrorCode.Should().Be(ErrorCodes.NoSelfArgument);
+            diagnostic.SourceSpan.Should().Be(4, 9, 4, 13);
+            diagnostic.Message.Should().Be(Resources.NoSelfArgument);
+        }
+
+        [TestMethod, Priority(0)]
+        public async Task FirstArgumentAbstractPropertyNotSelf() {
+            const string code = @"
+class Test:
+    @abstractproperty
+    def test(x, y, z):
+        pass
+";
+            var analysis = await GetAnalysisAsync(code);
+            analysis.Diagnostics.Should().HaveCount(1);
+
+            var diagnostic = analysis.Diagnostics.ElementAt(0);
+            diagnostic.ErrorCode.Should().Be(ErrorCodes.NoSelfArgument);
+            diagnostic.SourceSpan.Should().Be(4, 9, 4, 13);
+            diagnostic.Message.Should().Be(Resources.NoSelfArgument);
+        }
+
+
+        [TestMethod, Priority(0)]
+        public async Task FirstArgumentNotSelfMultiple() {
+            const string code = @"
+class Test:
+    def test(x, y, z):
+        pass
+
+    def test2(x, y, z):
+        pass
+";
+            var analysis = await GetAnalysisAsync(code);
+            analysis.Diagnostics.Should().HaveCount(2);
+
+            var diagnostic = analysis.Diagnostics.ElementAt(0);
+            diagnostic.ErrorCode.Should().Be(ErrorCodes.NoSelfArgument);
+            diagnostic.SourceSpan.Should().Be(3, 9, 3, 13);
+            diagnostic.Message.Should().Be(Resources.NoSelfArgument);
+
+            diagnostic = analysis.Diagnostics.ElementAt(1);
+            diagnostic.ErrorCode.Should().Be(ErrorCodes.NoSelfArgument);
+            diagnostic.SourceSpan.Should().Be(6, 9, 6, 14);
+            diagnostic.Message.Should().Be(Resources.NoSelfArgument);
+        }
+
+        [TestMethod, Priority(0)]
+        public async Task FirstArgumentClassMethod() {
+            const string code = @"
+class Test:
+    @classmethod
```

Comment:
What happens when the method has other decorators than `classmethod`, potentially in different orders?

Reply:
We process only the builtin decorators, see `PythonFunctionType.ProcessDecorators`.